lib/galaxy/model/__init__.py

```diff
@@ -1634,12 +1632,32 @@ def all_entry_points_configured(self):
             all_configured = ep.configured and all_configured
         return all_configured
 
-    def set_state(self, state):
+    def set_state(self, state: JobState) -> bool:
         """
-        Save state history
+        Save state history. Returns True if state has changed, else False.
         """
-        self.state = state
-        self.state_history.append(JobStateHistory(self))
+        if self.state == state:
+            # Nothing changed, no action needed
+            return False
+        session = object_session(self)
+        if session and self.id and state not in Job.finished_states:
+            # generate statement that will not revert DELETING or DELETED back to anything non-terminal
+            rval = session.execute(
+                update(Job.table)
+                .where(Job.table.c.id == self.id, ~Job.table.c.state.in_((Job.states.DELETING, Job.states.DELETED)))
```

Comment:
I think we are moving away from using `.table.c` for SA 2.0 , ping @jdavcs

Reply:
Thanks, updated that and the code I touched

Reply:
Yes: if `Foo` is a mapped class (not a subquery, etc.) and `bar` is a mapped attribute, then `Foo.table.c.bar` is equivalent to `Foo.bar`. So for the sake of simplicity, `Foo.bar`.